### PR TITLE
refactor amd render_kernel

### PR DESCRIPTION
**Pull request by ignaciosica**
### Refactor amd vector prefix and wmma rendering

**Changes:**
- no longer renders vectorized dtypes if not used (before they were hardcoded) (most process_replay diff is because of this)
- unify use of `half` and `_Float16` into ~~`_Float16`~~ `half` (the remaining process_replay diff is because of this)
- remove hardcoded dtype rendering in wmma kernel
- introduce `render_vector_prefix` method in renderer (suggested by Qazalin)

Can't test locally because I don't have and AMD setup but CI is green

process_replay run [here](https://github.com/ignaciosica/tinygrad/actions/runs/10490685490/job/29057988702)

### CODE REVIEW

File: tinygrad/renderer/cstyle.py [hunk 1]
```diff
@@ -382,11 +377,15 @@ class AMDRenderer(CStyleLanguage):
...
+  type_map = {dtypes.bfloat16: "hip_bfloat16", dtypes.half: "_Float16"}
+
+  def render_vector_prefix(self, dtype:DType) -> str:
+    vec, scal = self.render_dtype(dtype), self.render_dtype(dtype.scalar()),
+    return f"typedef {scal} {vec} __attribute__((ext_vector_type({dtype.count})));\nstatic inline __attribute__((device)) "+\
+           f"{vec} make_{vec}({', '.join([f'{scal} {x}' for x in _nms[:dtype.count]])}) {{ return {{ {', '.join(_nms[:dtype.count])} }}; }}"
 
   def render_kernel(self, function_name, kernel, bufs, uops, prefix=None) -> str:
     prefix = ["#define INFINITY (__builtin_inff())", "#define NAN (__builtin_nanf(\"\"))", "typedef long unsigned int size_t;"]
-    vec_dts = [("float", "float", 2), ("float", "float", 4), ("float", "float", 8), ("signed int", "int", 4), ("signed int", "int", 2)]
```
Comment (ignaciosica) [inline tinygrad/renderer/cstyle.py:389]: One concern I have is on the fixed vector dtypes of `signed int` and `int`.
I believe this does not render the base_type as `signed int` and rather use only `int` for the vector dtype

Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py:389]: Unfortunately none of the process_replay kernels use vectorized int so I can't tell if it works. Don't know if CI test it either

Comment (chenyuxyz) [reply tinygrad/renderer/cstyle.py:389]: you can check CI coverage by removing stuff and see if any test breaks. if not, can you add something that triggers these (likely in test_linearizer, construct a kernel and apply different upcast to it)

i pushed this to internal benchmark that runs on real AMD

Comment (Qazalin) [reply tinygrad/renderer/cstyle.py:389]: oh we only vectorize half and float now I think, there's no int4 after lowerer.

**Comment by Qazalin** [comment]
seeing slowdown in llama at https://github.com/tinygrad/tinygrad/commit/72a4b85950438637242db703bbdbc2867997acf1, might be compiler though:


interestingly no visible impact on the others.

File: tinygrad/renderer/cstyle.py (hunk 1 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py:389]: `test_dtype/test_upcast_ops` generates the following for `METAL` backend. Guess no vectorization for `int` dtypes, it just seems to unroll.

``` c
#include <metal_stdlib>
using namespace metal;
kernel void E_4n41(device int* data0, const device short* data1, const device int* data2, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  short val0 = *(data1+0);
  short val1 = *(data1+1);
  short val2 = *(data1+2);
  short val3 = *(data1+3);
  int val4 = *(data2+0);
  int val5 = *(data2+1);
  int val6 = *(data2+2);
  int val7 = *(data2+3);
  *(data0+0) = ((int)(val0)*val4);
  *(data0+1) = ((int)(val1)*val5);
  *(data0+2) = ((int)(val2)*val6);
  *(data0+3) = ((int)(val3)*val7);
}
# --- #
#include <metal_stdlib>
using namespace metal;
kernel void E_4n104(device unsigned int* data0, const device unsigned short* data1, const device unsigned int* data2, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  unsigned short val0 = *(data1+0);
  unsigned short val1 = *(data1+1);
  unsigned short val2 = *(data1+2);
  unsigned short val3 = *(data1+3);
  unsigned int val4 = *(data2+0);
  unsigned int val5 = *(data2+1);
  unsigned int val6 = *(data2+2);
  unsigned int val7 = *(data2+3);
  *(data0+0) = ((unsigned int)(val0)+val4);
  *(data0+1) = ((unsigned int)(val1)+val5);
  *(data0+2) = ((unsigned int)(val2)+val6);
  *(data0+3) = ((unsigned int)(val3)+val7);
}
```

Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py:389]: > i pushed this to internal benchmark that runs on real AMD

did it work?

**Comment by Qazalin** [comment]
also I kind of don't like half is _Float16 everywhere
image

can we keep the alias?

**Comment by ignaciosica** [comment]
> can we keep the alias?

Sure, adding back the alias

**Comment by ignaciosica** [comment]
> seeing slowdown in llama at https://github.com/tinygrad/tinygrad/commit/72a4b85950438637242db703bbdbc2867997acf1, might be compiler though:

Sorry I don't know how to handle that. I don't see how this changes can significantly affect performance.

**Comment by Qazalin** [comment]
I'm gonna push to benchmarks again and rerun with the compiler cache, let's see if it's a noop.

File: tinygrad/renderer/cstyle.py [hunk 2]
```diff
@@ -339,12 +339,12 @@ def get_kernel_modifier(self, uops:List[UOp]) -> str:
     # https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html
     return f"__launch_bounds__({maxThreadsPerBlock}) "
 
-code_for_op_hip = { UnaryOps.SQRT: lambda x,dtype: f"__ocml_sqrt_f{ {dtypes.half:16, dtypes.double:64}.get(dtype, 32)}({x})",
-                    UnaryOps.SIN: lambda x,dtype: f"__ocml_sin_f{ {dtypes.half:16, dtypes.double:64}.get(dtype, 32)}({x})",
-                    UnaryOps.LOG2: lambda x,dtype: f"__ocml_log2_f{ {dtypes.half:16, dtypes.double:64}.get(dtype, 32)}({x})",
-                    UnaryOps.EXP2: lambda x,dtype: f"__ocml_exp2_f{ {dtypes.half:16, dtypes.double:64}.get(dtype, 32)}({x})",
+code_for_op_hip = { UnaryOps.SQRT: lambda x, dtype: f"__ocml_sqrt_f{ {dtypes.half:16, dtypes.double:64}.get(dtype, 32) }({x})",
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:342]: ugh, why does spacing change...

Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: Sorry, I tried changing `{dtypes.half:16, dtypes.double:64}.get(dtype, 32)}` for `dtype.itemsize*8` but when I realized why it was wrong I revert the changes and missed reverting spaces too

**Comment by Qazalin** [comment]
I want you to honestly ask yourself if you read the diff very carefully. If you wanna get your PRs merged fast you need to read your diff 3 times before marking as "ready for review". This builds trust with the repo maintainers.

That WMMA change isn't even generic to support other dtypes (it shouldn't...), idk why you're (half) removing the hardcoded fp16 WMMA.

**Comment by ignaciosica** [comment]
I changed wmma dtype rendering to support the unification of `_Float16` and `half` and removal of the alias. Once you told me to add the alias back I thought there was no harm in leaving that diff on wmma rendering. I can revert if you want

File: tinygrad/renderer/cstyle.py (hunk 2 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: And about the spacing is the same, I didn't really missed it, when I was reverting the changes I thought might as well leave the fix for spacing, I found it a little inconsistent across the codebase. But now I know that for a separate PR.

image

**Comment by ignaciosica** [comment]
Reverting based on chenyu's discord comment but I do believe it was more correct the other way.

**Pull request merged**